From ba7deefcab6f06f6dd66bdfc5ef03e1bde7cfdd1 Mon Sep 17 00:00:00 2001
From: Dan Lambright <dlambrig@redhat.com>
Date: Thu, 1 Oct 2015 11:05:25 -0400
Subject: [PATCH 104/131] cluster/tier: Disallow detach commit when detach in progress

1. Check if detach is running, disallow detach commit if so.
2. Cleanup shutdown of tier daemon on detach: do not rerun fix-layout,
   do not send incorrect status back to glusterd.

> Change-Id: I97202f748773c1176396a4ffd32a4c7fa9b9c1bc
> BUG: 1279637
> Signed-off-by: Dan Lambright <dlambrig@redhat.com>
> Reviewed-on: http://review.gluster.org/12560
> Tested-by: Gluster Build System <jenkins@build.gluster.com>
> Reviewed-by: Atin Mukherjee <amukherj@redhat.com>
Signed-off-by: Dan Lambright <dlambrig@redhat.com>

Signed-off-by: Dan Lambright <dlambrig@redhat.com>

Conflicts:
	xlators/cluster/dht/src/tier.c

Change-Id: I4d9617a6ca2b8b298ff80fd343cea82da9192afb
BUG: 1278390
Reviewed-on: https://code.engineering.redhat.com/gerrit/61589
Reviewed-by: Dan Lambright <dlambrig@redhat.com>
Tested-by: Dan Lambright <dlambrig@redhat.com>
---
 xlators/cluster/dht/src/dht-rebalance.c        |    3 --
 xlators/mgmt/glusterd/src/glusterd-brick-ops.c |   25 ++++++++++++++++++++++++
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/xlators/cluster/dht/src/dht-rebalance.c b/xlators/cluster/dht/src/dht-rebalance.c
index 5e701b3..6c9206b 100644
--- a/xlators/cluster/dht/src/dht-rebalance.c
+++ b/xlators/cluster/dht/src/dht-rebalance.c
@@ -3378,9 +3378,6 @@ gf_defrag_start_crawl (void *data)
                         if (ret)
                                 goto out;
 
-                        ret = gf_defrag_fix_layout (this, defrag, &loc,
-                                                    fix_layout,
-                                                    migrate_data);
                 }
         }
         gf_log ("DHT", GF_LOG_INFO, "crawling file-system completed");
diff --git a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
index 3ca5dcf..7380356 100644
--- a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
+++ b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
@@ -2040,6 +2040,31 @@ glusterd_op_stage_remove_brick (dict_t *dict, char **op_errstr)
                                 GD_MSG_VOL_NOT_TIER, "%s", errstr);
                         goto out;
                 }
+                ret = glusterd_remove_brick_validate_bricks (cmd, brick_count,
+                                                             dict, volinfo,
+                                                             &errstr);
+                if (ret)
+                        goto out;
+
+                /* If geo-rep is configured, for this volume, it should be
+                 * stopped.
+                 */
+                param.volinfo = volinfo;
+                ret = glusterd_check_geo_rep_running (&param, op_errstr);
+                if (ret || param.is_active) {
+                        ret = -1;
+                        goto out;
+                }
+
+                if (volinfo->rebal.defrag_status == GF_DEFRAG_STATUS_STARTED) {
+                        ret = -1;
+                        errstr = gf_strdup("Detach is in progress. Please "
+                                           "retry after completion");
+                        gf_msg (this->name, GF_LOG_ERROR, 0,
+                                GD_MSG_OIP_RETRY_LATER, "%s", errstr);
+                        goto out;
+                }
+                break;
 
         case GF_OP_CMD_COMMIT:
                 if (volinfo->decommission_in_progress) {
-- 
1.7.1

